update pointer bitcast management

rjodinchr opened the pull request:
- clean GetIdxsForTyFromOffset for Constant Path with structure types
- reorganise simplify function
- add runOnUnneededIndices to avoid implicitCasts issue with #1109
- add debug log in simplify pointer bitcast pass
- update & add tests

This PR is needed to solve issue from #1109

Fix #1115

alan-baker commented on test/PointerCasts/opaque_trivial_casts.ll:
```diff
@@ -78,17 +78,16 @@ entry:
 ; CHECK: }
 
-define void @test6(ptr %in) {
+define void @test6(ptr addrspace(1) %in) {
```
I think this is the same as test7 now.

alan-baker reviewed (commented):
I'm not sure how this is a fix for #1115. The test simply points out that we generate incorrect code. Changing the test to a different address space doesn't change that fact.

rjodinchr commented:
You are right, I did not notice that it was the same as `test7`.
I have removed `test6` and added a test showing how we should output an error when we encounter such invalid cast.
